src/io/checkpoint_io.f90

```diff
@@ -633,139 +722,266 @@ subroutine restart_checkpoint( &
+        "v", field_data_v, isolated_file, start_dims, count_dims)
+      call isolated_reader%close(isolated_file)
+      call isolated_reader%finalise()
+
+      call isolated_reader%init(comm, "w_reader")
+      isolated_file = isolated_reader%open(filename, adios2_mode_read, comm)
+      call isolated_reader%begin_step(isolated_file)
+      call isolated_reader%read_data( &
+        "w", field_data_w, isolated_file, start_dims, count_dims)
+      call isolated_reader%close(isolated_file)
+      call isolated_reader%finalise()
+
+      do i = 1, size(field_names)
+        select case (trim(field_names(i)))
+        case ("u")
+          current_field_data => field_data_u
+        case ("v")
+          current_field_data => field_data_v
+        case ("w")
+          current_field_data => field_data_w
+        end select
+
+        select case (trim(field_names(i)))
+        case ("u")
+          call solver%backend%set_field_data(solver%u, current_field_data)
+        case ("v")
+          call solver%backend%set_field_data(solver%v, current_field_data)
+        case ("w")
+          call solver%backend%set_field_data(solver%w, current_field_data)
+        case default
+          call self%adios2_writer%handle_error( &
+            1, "Invalid field name"//trim(field_names(i)))
+        end select
+      end do
+      deallocate (field_data_u, field_data_v, field_data_w)
+    end block
 
     call reader%close(file)
     call reader%finalise()
-    if (allocated(field_data)) deallocate (field_data)
   end subroutine restart_checkpoint
 
-  subroutine write_fields(self, field_names, fields, solver, file, use_stride)
-    !! Write field data, optionally with striding
+  subroutine write_fields( &
+    self, field_names, fields, host_fields, solver, file, &
+    use_stride, writer &
+    )
+    !! Write field data, optionally with striding, using separate buffers for true async I/O
     class(checkpoint_manager_adios2_t), intent(inout) :: self
     character(len=*), dimension(:), intent(in) :: field_names
-    class(field_ptr_t), dimension(:), target, intent(in) :: fields
+    class(field_ptr_t), dimension(:), target, intent(in) :: fields, host_fields
     class(solver_t), intent(in) :: solver
     type(adios2_file_t), intent(inout) :: file
     logical, intent(in), optional :: use_stride
+    type(adios2_writer_t), intent(inout) :: writer
 
-    class(field_t), pointer :: host_field
     integer :: i_field
     logical :: apply_stride
 
-    call self%cleanup_strided_buffers()
-
     apply_stride = .false.
     if (present(use_stride)) apply_stride = use_stride
 
+    ! pre-allocate separate buffers for each field to enable race-free async I/O
+    if (size(field_names) >= 3) then
+      if (apply_stride) then
+        call prepare_field_buffers(solver, self%output_stride, field_names)
+      else
+        call prepare_field_buffers( &
+          solver, self%checkpoint_stride_factors, field_names)
+      end if
```

Comment:
This logic doesn't seem quite right, or I'd rename to `use_output/checkpoint_stride` for the test